### ISSUE TITLE
[BUG] RKE2 Downstream Clusters Not Coming Up After Rancher Migration

### ISSUE THREAD

**Issue by nickwsuse** [rancher/rancher#40080]
**Rancher Server Setup**
- Rancher version: v2.6.9 && v2.7.0
- Installation option (Docker install/Helm Chart): Helm
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): RKE2 
- Proxy/Cert Details: byo-valid

**Information about the Cluster**
- Kubernetes version: v1.24.8+rke2r1
- Cluster Type (Local/Downstream): Downstream 
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): AWS

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom): Admin

**Describe the bug**
When migrating rancher servers from one HA to another HA using Rancher Backups and Restore, the RKE2 downstream cluster is not coming back up. The status is stuck at `Updating` with the following message: `Configuring bootstrap node(s) <redacted>: waiting for plan to be applied`

The RKE2 version I used for both Rancher versions is `v1.24.8+rke2r1` and after changing the version to the default RKE2 version for each Rancher version (`v1.24.4+rke2r1` && `v1.24.6+rke2r1` respectively) and redeploying all workloads, the RKE2 cluster for the v2.6.9 instance came up and was `Active`. 

The v2.7.0 RKE2 cluster is still showing the same status and message.

**To Reproduce**

1. Deploy a Rancher HA instance on v2.6.9 && v2.7.0
2. Create an AWS RKE1 downstream cluster (3 workers, 1 control plane, 1 etcd) using `v1.24.8` as the RKE version
3. Create an AWS RKE2 downstream cluster (3 workers, 2 control plane, 3 etcd) using `v1.24.8+rke2r1` as the RKE version
4. Wait for the downstream clusters to be `Active`
5. Install the Rancher Backups chart on your local cluster
6. Create a backup in your preferred storage (I use an AWS S3 bucket)
7. Bring up a new HA and point the load balancer to the new HA
8. Use the backup to restore onto the new HA
9. Install Rancher with the same version as the original HA
10. Go to Cluster Management
11. Check cluster statuses

**Result**
The RKE2 downstream cluster did **not** come back to the `Active` status after the restore/migration

**Expected Result**

The RKE2 downstream cluster does come back to the `Active` after the restore/migration

**Screenshots**

Showing the Status and Message:
![image](https://user-images.githubusercontent.com/107147272/210901860-1bc3586a-bb52-4843-a9cb-481a1e6115b3.png)

Machine Pool:
![image](https://user-images.githubusercontent.com/107147272/210902096-ca440a8b-13c6-4fcc-8bec-cdabb4237b2a.png)

**Comment by nickwsuse** [comment]
I provisioned some downstream clusters using k8s v1.23.14-rancher1-1 and had this same issue occur again, so it doesn't seem to be tied to the k8s version from what I can see.

**Comment by hwaastad** [comment]
I can verify that it doesnt seem to be tied to the k8s version.
I’ve tester from 1.20.x and up (rke 1.4.1 for admin HA cluster)

**Comment by hwaastad** [comment]
@nickwsuse after digging a bit more it seems that after migration, the custom-xxxx-machine-plan secret is missing after restore. As far as I can see this is the token being used by the rancher-system-agent and has ownerref to the custom-xxxx-machine-plan SA.
On my second try, I added the custom-xxxx-machine-plan secret in "secrets" on the custom-xxxx-machine-plan ServiceAccount before taking backup. Now the restore worked. If this make sense or not I do not know. If needed I can do more testing.

**Comment by eliyamlevy** [comment]
Looking into this now 👍

**Comment by hwaastad** [comment]
@eliyamlevy, if you find any workaround before any fix, please share.
I’ve tried to modify the default resourceset to include the machine-plan-token$ but it does’nt seem to work.

**Comment by MKlimuszka** [comment]
This will need to be fixed in both 2.6x and 2.7x

**Comment by rancherbot** [comment]
@eliyamlevy, Not creating `backport` issue for issue `40080` in repository `rancher` because milestone `2.6.10` does not exist or is not an open milestone

**Comment by rancherbot** [comment]
@eliyamlevy, Not creating `backport` issue for issue `40080` in repository `rancher` because milestone `v2.6.10` does not exist or is not an open milestone

**Comment by harridu** [comment]
2 questions:

- Is this a bug on backup, or on restore? Are the custom-xxxx-machine-plan secrets included in the backup file?
- Is there a workaround to repair damaged Rancher clusters?

**Comment by nickwsuse** [comment]
UPDATE:

The fix that was implemented didn't appear to work, and upon further investigation this seems to be a bug in the rancher-system-agent.

It seems like anything that leads to rancher-system-agent being invoked to apply a change to the nodes has a high chance of hitting this. Doesn’t matter if it’s a restore or just a minor config change. 

Currently waiting on the correct engineering team to start looking into it.

**Comment by snasovich** [comment]
@eliyamlevy , duplicating offline conversation, please rollback https://github.com/rancher/backup-restore-operator/pull/293 as we don't want to backup `*machine-plan-token` secrets (or any secrets of `kubernetes.io/service-account-token` type) as these are service account token that won't be valid if restored on a new cluster as they are tied to service account ID that will be different on new cluster.

**Comment by Oats87** [comment]
I was able to successfully reproduce this issue. The root cause stems from the fact that the system-agent relies on a service account token to communicate with Rancher, and the corresponding Kubernetes cluster underneath Rancher. This service account token is generated when the system-agent installation is run, i.e. system-agent-install.sh is run for a custom cluster, via an SSH session, or for a provisioned cluster, via cloud-init. It can also be run by the system-upgrade-controller when it is attempting to manage the system-agent on the node. When a backup-restore operation occurs, the service account token that is stored for the system-agent on every node is invalidated (due to the changing base cluster), and the system-agent is unable to communicate with Rancher.

The simplest way to resolve this is by running the system-upgrade-controller based `system-agent-upgrader` jobs on the nodes in the cluster. A workaround for this is to remove the label key `plan.upgrade.cattle.io/system-agent-upgrader` from all of the downstream nodes. This will cause the system-upgrade-controller to run the `apply-system-agent-upgrader` jobs on the nodes, which will invoke the `system-agent-installer` and generate a new service account token for the node.

There appears to be another issue with the `system-upgrade-controller` itself causes a rerun of the various jobs, but in our case, this bug seems to actually be beneficial for bringing a cluster back into communication with Rancher.

I'm not so sure I would call this a regression -- I believe this has likely been occurring from day 0 of Rancher v2prov with the backup-restore operator.

**Comment by Oats87** [comment]
Okay, update to this.

It turns out our issue is a little deeper than originally determined.

The rancher-system-agent is force-upgraded (through a corresponding fleet bundle) in the downstream cluster when the rancher backup operator restores the cluster. This is done via a generation field that is contained within the spec of the v1 provisioning cluster, that then trickles a change into the fleet bundle that is used for managing the system agent plan for system-upgrade-controller.

We are seeing an intermittent issue (most likely a race) where the corresponding Rancher v3 token value and the kubeconfig secret (for CAPI) that is generated on the fly do not match, which causes fleet to be unable to update the credentials for fleet-agent in the downstream cluster (because we are dealing with a new base cluster). Because of this, fleet-agent is unable to reconcile the updated bundle for the SUC system-agent plan and thus we end up in our case where system-agent is not updated because the SUC plan never runs, which never generates new credentials for the system-agent. This then causes Rancher to sit forever waiting for the nodes to report plan application, which never actually happens because the nodes do not have correct credentials to talk to rancher to retrieve the latest plan.

Further investigation is going to be required, as this is seemingly a bug that has probably existed for a while.

It is easy to reproduce, if you scale rancher to 0 replicas, and delete the corresponding cluster kubeconfig (in the `fleet-default` namespace), and the v3 user token, then scale rancher back up, there is a 30% chance of having Rancher race and generate an invalid kubeconfig secret.

**Comment by eliyamlevy** [comment]
> @eliyamlevy , duplicating offline conversation, please rollback [rancher/backup-restore-operator#293](https://github.com/rancher/backup-restore-operator/pull/293) as we don't want to backup `*machine-plan-token` secrets (or any secrets of `kubernetes.io/service-account-token` type) as these are service account token that won't be valid if restored on a new cluster as they are tied to service account ID that will be different on new cluster.

@snasovich reverted in most recent rcs for 2.6 and 2.7

**Comment by snasovich** [comment]
This should be addressed by the same fix done for https://github.com/rancher/rancher/issues/40741, that's why it's moved "To Test" at the same time.

**Comment by nickwsuse** [comment]
I'm sort of blocked on testing this until https://github.com/rancher/rancher/issues/40819 is resolved, as my RKE2 clusters are stuck `Updating` due to the backup/restore issue. Once that issue is fixed I'll try again and make sure the cluster comes back to the `Active` status.
